lib/ide/utils/src/handle.rs

```diff
@@ -0,0 +1,268 @@
+//! This module implements utilities that allow for a safer usage pattern of
+//! a shared data stored under `RefCell`. The primary goal is to allow accessing
+//! the data without any direct calls to `borrow_mut`.
+//!
+//! Any type that implements `IsWeakHandle` or `IsStrongHandle` (types provided
+//! out-of-the-box are Weak<RefCell<T>> and Strong<RefCell<T>> responsively)
+//! gets an extension method names `with`. It allows accessing a data through
+//! callback function that will get &mut access to the `T`.
+//!
+//! The safety is ensured by the execution model. The `with` function returns
+//! a `Future` that yields a result of callback. Because `borrow_mut` only
+//! applied in the `Future` implementation (being called by the executor), it
+//! is guaranteed to be safe (as executors do not allow self-entry and callback
+//! must be a synchronous function).
+
+use crate::prelude::*;
+
+use std::future::Future;
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+/// Packs given value into a Rc<RefCell<T>>.
+pub fn strong<T>(t:T) -> StrongHandle<T> { Rc::new(RefCell::new(t)) }
+
+/// A strong (shared ownership) handle to some internally mutable data.
+pub type StrongHandle<T> = Rc<RefCell<T>>;
+
+/// A weak handle to some internally mutable data.
+pub type WeakHandle<T> = Weak<RefCell<T>>;
+
+/// Value or error if the data under handle was not available.
+pub type Result<T> = std::result::Result<T,Error>;
+
+
+
+// ===========
+// == Error ==
+// ===========
+
+/// Error that may occur when trying to access data through a `WeakHandle`.
+#[derive(Clone,Copy,Display,Debug,Fail)]
+pub enum Error {
+    /// Happens when `Weak` cannot be upgraded.
+    /// Likely the object (if there was an object) has been already dropped.
+    HandleExpired,
+    /// RefCell was already borrowed, cannot `borrow_mut`. Should never happen.
+    AlreadyBorrowed,
+}
+
+
+
+// ================
+// == WeakHandle ==
+// ================
+
+/// A type that can provide `Weak` handle to a value under `RefCell`.
+pub trait IsWeakHandle: Clone {
+    /// Type of the data stored.
+    type Data;
```

Comment:
This is a bad pattern. Basically the most composable way to do it is to create a separate trait only for type dependencies. An example of what it leads to is here. Right now I can have a polymorphic function `foo<T>(t:T)` and I want it to return its "Data" no matter if it was weak or strong handle. I cannot do it simply. Instead you should do:

```rust
pub trait HasContent {
    type Content;
}

pub trait IsWeakHandle : Clone + HasContent { ... }
```

And FYI, HasContent is already defined in Prelude :)

Moreover, the implementations would so simplify using this pattern . -we can then use default impls in an easy way.

Reply:
Could you provide better example than `foo<T>(t:T)`
I cannot figure out what can be an example of function which operate on trait with near-to-none semantic value.

Reply:
Superseded by other suggested changes, there are no traits anymore.